domain/port/state/watcher.go

```diff
@@ -0,0 +1,107 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+
+	"github.com/canonical/sqlair"
+	jujuerrors "github.com/juju/errors"
+
+	"github.com/juju/juju/internal/errors"
+)
+
+// WatchOpenedPortsTable returns the name of the table that should be watched
+func (st *State) WatchOpenedPortsTable() string {
+	return "port_range"
+}
+
+// InitialWatchOpenedPortsStatement returns the query to load the initial event
+// for the WatchOpenedPorts watcher
+func (st *State) InitialWatchOpenedPortsStatement() string {
+	return "SELECT uuid FROM machine"
+}
+
+// GetMachinesForEndpoints returns map from endpoint uuids to the uuids of the
+// machines which host that endpoint for each provided endpoint uuid.
+func (st *State) GetMachinesForEndpoints(ctx context.Context, eps []string) (map[string]string, error) {
+	db, err := st.DB()
+	if err != nil {
+		return nil, jujuerrors.Trace(err)
+	}
+
+	endpointUUIDs := endpoints(eps)
+
+	query, err := st.Prepare(`
+SELECT
+	machine.uuid AS &endpointAndMachineUUID.machine_uuid,
```

Comment:
This is not the agreed-upon formatting.

Reply:
This was fixed as a consequence of the fix to the above